hack/role/role.go

```diff
@@ -0,0 +1,380 @@
+			"0000_50_cluster-ingress-operator_00-ingress-credentials-request.yaml",
+			"0336e1d3-7a87-462b-b6db-342b63f7802c",
+		},
+		{
+			"Cluster Network Operator",
+			"0000_50_cluster-network-operator_02-cncc-credentials.yaml",
+			"be7a6435-15ae-4171-8f30-4a343eff9e8f",
+		},
+		{
+			"Cluster Storage Operator",
+			"0000_50_cluster-storage-operator_03_credentials_request_azure.yaml",
+			"5b7237c5-45e1-49d6-bc18-a1f62f400748",
+		},
+		{
+			"Cluster Storage Operator (Azure File)",
+			"0000_50_cluster-storage-operator_03_credentials_request_azure_file.yaml",
+			"0d7aedc0-15fd-4a67-a412-efad370c947e",
+		},
+	}
+)
+
+func main() {
+	flag.Parse()
+	if *verifiedVersion == "" {
+		panic("verified-version is required")
+	}
+
+	ctx := context.Background()
+
+	logger := logrus.New()
+	logger.SetLevel(logrus.ErrorLevel)
+	log := logrus.NewEntry(logger)
+	environment, err := env.NewCoreForCI(ctx, log) // we don't use log here
+	if err != nil {
+		panic(err)
+	}
+
+	options := environment.Environment().EnvironmentCredentialOptions()
+	tokenCredential, err := azidentity.NewEnvironmentCredential(options)
+	if err != nil {
+		panic(err)
+	}
+	roleDefinitions, err := armauthorization.NewArmRoleDefinitionsClient(tokenCredential, environment.SubscriptionID(), nil)
+	if err != nil {
+		panic(err)
+	}
+
+	m := manager{
+		roleDefinitions: roleDefinitions,
+	}
+
+	verifiedDir, err := os.MkdirTemp("", *verifiedVersion)
+	if err != nil {
+		panic(err)
+	}
+	err = extractCredReq(*verifiedVersion, verifiedDir)
+	if err != nil {
+		panic(err)
+	}
+
+	var vers []string
+	if *targetVersion != "" {
+		// if target version is specified, validate only that version
+		vers = []string{*targetVersion}
+	} else {
+		// validate all available versions
+		vers, err = versionsToValidate(*verifiedVersion)
+		if err != nil {
+			panic(err)
+		}
+	}
+
+	var missing []string
+	for _, v := range vers {
+		fmt.Println("Checking", v)
+		targetDir, err := os.MkdirTemp("", v)
+		if err != nil {
+			panic(err)
+		}
+		err = extractCredReq(v, targetDir)
+		if err != nil {
+			panic(err)
+		}
+
+		if err = validate(ctx, &m, verifiedDir, targetDir, permissionSets); err != nil {
+			// To check all versions, we need to continue even if there are missing permissions.
+			if errors.Is(err, &missingPermissionsError{}) {
+				missing = append(missing, v)
+				continue
+			}
+			panic(err)
+		}
+	}
+	if len(missing) > 0 {
+		for _, v := range missing {
+			fmt.Printf("Permissions are missing in %s\n", v)
+		}
+		os.Exit(1)
+	}
+}
+
+// versionsToValidate returns available versions later than the verified version
+func versionsToValidate(verifiedVersion string) ([]string, error) {
+	v, err := version.ParseVersion(verifiedVersion)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse version: %w", err)
+	}
+
+	var vers []string
+	for ; ; v.V[1]++ {
+		res, err := http.Get(fmt.Sprintf("https://api.openshift.com/api/upgrades_info/v1/graph?channel=candidate-%s", v.MinorVersion()))
```

Comment:
One potential idea - rather than looking at the upgrade graph, can the automation just compare the latest OCP build (right now, it's 4.18.0-ec.2-x86_64) with the verified-version provided to the command? We will always want to make sure our roles match the permissions extracted from the latest build.

Reply:
Actually candidate channel has the rc/ec versions, but I decided to chang to use quay.io API (see the commit for details).
The only concern about rc/ec was that it's impossible to tell which is newer just based on the version.
By using quay.io, we can use last_modified date time to determine which is the latest.
Although it doesn't represent the exact release date, it's safe to assume that last modified one should be the latest version because they're pushed by CI.

This is the local output.

<details><summary>Details</summary>
<p>

```
❯ OCP_VERSION=4.14.37 make validate-roledef
go run ./hack/role -verified-version "4.14.37" -oc-bin=oc
Extracting 4.14.37
Checking 4.14.38
Extracting 4.14.38
Validating permissions
Checking 4.15.36
Extracting 4.15.36
Validating permissions
        MachineApiOperator: missing actions:
                Microsoft.Network/virtualNetworks/join/action
        ImageRegistryOperator: missing actions:
                Microsoft.Network/privateEndpoints/write
                Microsoft.Network/privateEndpoints/read
                Microsoft.Network/privateEndpoints/privateDnsZoneGroups/write
                Microsoft.Network/privateEndpoints/privateDnsZoneGroups/read
                Microsoft.Network/privateDnsZones/read
                Microsoft.Network/privateDnsZones/write
                Microsoft.Network/privateDnsZones/join/action
                Microsoft.Network/privateDnsZones/A/write
                Microsoft.Network/privateDnsZones/virtualNetworkLinks/write
                Microsoft.Network/privateDnsZones/virtualNetworkLinks/read
                Microsoft.Network/networkInterfaces/read
                Microsoft.Storage/storageAccounts/PrivateEndpointConnectionsApproval/action
                Microsoft.Network/virtualNetworks/subnets/read
                Microsoft.Network/virtualNetworks/subnets/join/action
                Microsoft.Network/virtualNetworks/join/action
Checking 4.16.17
Extracting 4.16.17
Validating permissions
        MachineApiOperator: missing actions:
                Microsoft.Network/virtualNetworks/join/action
        ImageRegistryOperator: missing actions:
                Microsoft.Network/privateEndpoints/write
                Microsoft.Network/privateEndpoints/read
                Microsoft.Network/privateEndpoints/privateDnsZoneGroups/write
                Microsoft.Network/privateEndpoints/privateDnsZoneGroups/read
                Microsoft.Network/privateDnsZones/read
                Microsoft.Network/privateDnsZones/write
                Microsoft.Network/privateDnsZones/join/action
                Microsoft.Network/privateDnsZones/A/write
                Microsoft.Network/privateDnsZones/virtualNetworkLinks/write
                Microsoft.Network/privateDnsZones/virtualNetworkLinks/read
                Microsoft.Network/networkInterfaces/read
                Microsoft.Storage/storageAccounts/PrivateEndpointConnectionsApproval/action
                Microsoft.Network/virtualNetworks/subnets/read
                Microsoft.Network/virtualNetworks/subnets/join/action
                Microsoft.Network/virtualNetworks/join/action
        ClusterIngressOperator: missing actions:
                Microsoft.Network/virtualNetworks/subnets/read
                Microsoft.Network/virtualNetworks/subnets/join/action
Checking 4.17.1
Extracting 4.17.1
Validating permissions
        MachineApiOperator: missing actions:
                Microsoft.Network/virtualNetworks/join/action
        ImageRegistryOperator: missing actions:
                Microsoft.Storage/storageAccounts/blobServices/containers/delete
                Microsoft.Network/privateEndpoints/write
                Microsoft.Network/privateEndpoints/read
                Microsoft.Network/privateEndpoints/privateDnsZoneGroups/write
                Microsoft.Network/privateEndpoints/privateDnsZoneGroups/read
                Microsoft.Network/privateDnsZones/read
                Microsoft.Network/privateDnsZones/write
                Microsoft.Network/privateDnsZones/join/action
                Microsoft.Network/privateDnsZones/A/write
                Microsoft.Network/privateDnsZones/virtualNetworkLinks/write
                Microsoft.Network/privateDnsZones/virtualNetworkLinks/read
                Microsoft.Network/networkInterfaces/read
                Microsoft.Storage/storageAccounts/PrivateEndpointConnectionsApproval/action
                Microsoft.Network/virtualNetworks/subnets/read
                Microsoft.Network/virtualNetworks/subnets/join/action
                Microsoft.Network/virtualNetworks/join/action
        ClusterIngressOperator: missing actions:
                Microsoft.Network/virtualNetworks/subnets/read
                Microsoft.Network/virtualNetworks/subnets/join/action
        AzureFilesStorageOperator: missing actions:
                Microsoft.Network/routeTables/join/action
Checking 4.18.0-ec.2
Extracting 4.18.0-ec.2
Validating permissions
        MachineApiOperator: missing actions:
                Microsoft.Network/virtualNetworks/join/action
        ImageRegistryOperator: missing actions:
                Microsoft.Storage/storageAccounts/blobServices/containers/delete
                Microsoft.Network/privateEndpoints/write
                Microsoft.Network/privateEndpoints/read
                Microsoft.Network/privateEndpoints/privateDnsZoneGroups/write
                Microsoft.Network/privateEndpoints/privateDnsZoneGroups/read
                Microsoft.Network/privateDnsZones/read
                Microsoft.Network/privateDnsZones/write
                Microsoft.Network/privateDnsZones/join/action
                Microsoft.Network/privateDnsZones/A/write
                Microsoft.Network/privateDnsZones/virtualNetworkLinks/write
                Microsoft.Network/privateDnsZones/virtualNetworkLinks/read
                Microsoft.Network/networkInterfaces/read
                Microsoft.Storage/storageAccounts/PrivateEndpointConnectionsApproval/action
                Microsoft.Network/virtualNetworks/subnets/read
                Microsoft.Network/virtualNetworks/subnets/join/action
                Microsoft.Network/virtualNetworks/join/action
        ClusterIngressOperator: missing actions:
                Microsoft.Network/virtualNetworks/subnets/read
                Microsoft.Network/virtualNetworks/subnets/join/action
        AzureFilesStorageOperator: missing actions:
                Microsoft.Network/routeTables/join/action
Permissions are missing in 4.15.36
Permissions are missing in 4.16.17
Permissions are missing in 4.17.1
Permissions are missing in 4.18.0-ec.2
exit status 1
make: *** [validate-roledef] Error 1
```

</p>
</details>